### PR TITLE
Selective mode play

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
An extension of the current play tool by supporting selectively saving only part of the episodes.
This is useful in several cases, e.g. discovering failure modes.

### CODE REVIEW

File: alf/trainers/policy_trainer.py [hunk 1]
```diff
@@ -743,16 +743,69 @@ def _step(algorithm,
...
+        metrics (StepMetric): a list of metrics that will be updated based on
+            ``time_step``.
+        render (bool|False): if True, display the frames of ``env`` on a screen.
+        recorder (VideoRecorder|None): recorder the frames of ``env`` and other
+            additional images in prediction step info if present.
+        sleep_time_per_step (int|0): The sleep time between two frames when
+            ``render`` is True.
+        selective_criteria_func (callable|None): a callable for determining
+            whether an episode will be saved to the video file when a valid
+            recorder is provided.
```
Comment (emailweixu) [inline alf/trainers/policy_trainer.py]: should document its arguments

File: alf/trainers/policy_trainer.py [hunk 2]
```diff
@@ -743,16 +743,69 @@ def _step(algorithm,
...
+    if recorder and selective_criteria_func is None:
         recorder.capture_frame(policy_step.info, time_step.is_last())
+
+    elif recorder and selective_criteria_func is not None:
+        env_frame = recorder.capture_env_frame()
+        recorder.cache_frame_and_pred_info(env_frame, policy_step.info)
+
+        if time_step.is_last():
+            # below is an example failure_criteria based on return.
+            # This should be adjusted according to the particular task at hand.
```
Comment (emailweixu) [inline alf/trainers/policy_trainer.py]: Should be removed. They are at 934-935

File: alf/trainers/policy_trainer.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/trainers/policy_trainer.py]: Now added documents on the two arguments.

File: alf/trainers/policy_trainer.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/trainers/policy_trainer.py]: Removed.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**